internal/controller/garbagecollector_controller.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"time"
+
+	velerov1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	nacv1alpha1 "github.com/migtools/oadp-non-admin/api/v1alpha1"
+	"github.com/migtools/oadp-non-admin/internal/common/constant"
+	"github.com/migtools/oadp-non-admin/internal/common/function"
+	"github.com/migtools/oadp-non-admin/internal/predicate"
+)
+
+// GarbageCollectorReconciler reconciles Velero objects
+type GarbageCollectorReconciler struct {
+	client.Client
+	Scheme        *runtime.Scheme
+	OADPNamespace string
+	Frequency     time.Duration
+}
+
+var previousGarbageCollectorRun *time.Time
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *GarbageCollectorReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	labelSelector := client.MatchingLabels{
+		constant.OadpLabel:      constant.OadpLabelValue,
+		constant.ManagedByLabel: constant.ManagedByLabelValue,
+	}
+
+	if previousGarbageCollectorRun == nil || time.Now().After(previousGarbageCollectorRun.Add(r.Frequency)) {
+		previousGarbageCollectorRun = ptr.To(time.Now())
+		logger.V(1).Info("Garbage Collector Reconcile start")
+		// TODO duplication in delete logic
+		// TODO do deletion in parallel?
+
+		// TODO delete secret as well?
```

Comment:
should also check for possible leftover secrets?

Reply:
Correct, those should be cleaned up by the NABSL controller. On top of that now when I think about that we may want to add OwnerReference between Secret <-> Velero BSL object. This will cause Secret to be removed if the Velero BSL object is gone, so we don't have to implement this part in GC coontroller ?

Reply:
that would be good

but secret is created prior to bsl, can be added independent of the creation order?

Reply:
You are right. We can not add or patch the *ownerReference* on the Secret which already exists. What we could do is to add it vice-versa to the Velero BSL and make Velero BSL having *ownerReference* on the secret and then add `blockOwnerDeletion` on the Secret object. This should prevent secret to be removed independently from the BSL, but as always it's not 100% guarantee and there are edge cases where this still may happen.